lib/shrine.rb

```diff
@@ -29,966 +31,325 @@ def initialize(io, missing_methods)
-          self.id == other.id &&
-          self.storage_key == other.storage_key
-        end
-        alias eql? ==
+    # Calls `#upload` on the storage, passing to it the location, metadata
+    # and any upload options. The storage might modify the location or
+    # metadata that were passed in. The uploaded IO is then closed.
+    def copy(io, context)
+      location       = context[:location]
+      metadata       = context[:metadata]
+      upload_options = context[:upload_options] || {}
+
+      storage.upload(io, location, shrine_metadata: metadata, **upload_options)
+    ensure
+      io.close rescue nil
+    end
 
-        # Enables using UploadedFile objects as hash keys.
-        def hash
-          [id, storage_key].hash
-        end
+    # Delegates to `UploadedFile#delete`.
+    def remove(uploaded_file, context)
+      uploaded_file.delete
+    end
 
-        # Returns an uploader object for the corresponding storage.
-        def uploader
-          shrine_class.new(storage_key)
-        end
+    # Delegates to #process.
+    def processed(io, context)
+      process(io, context)
+    end
 
-        # Returns the storage that this file was uploaded to.
-        def storage
-          shrine_class.find_storage(storage_key)
-        end
+    # Retrieves the location for the given IO and context. First it looks
+    # for the `:location` option, otherwise it calls #generate_location.
+    def get_location(io, context)
+      location = context[:location] || generate_location(io, context)
+      location or raise Error, "location generated for #{io.inspect} was nil (context = #{context})"
+    end
 
-        # Returns the Shrine class that this file's class is namespaced under.
-        def shrine_class
-          self.class.shrine_class
-        end
+    # If the IO object is a Shrine::UploadedFile, it simply copies over its
+    # metadata, otherwise it calls #extract_metadata.
+    def get_metadata(io, context)
+      if io.is_a?(UploadedFile)
+        io.metadata.dup
+      else
+        extract_metadata(io, context)
+      end
+    end
 
-        private
+    # Asserts that the object is a valid IO object, specifically that it
+    # responds to `#read`, `#eof?`, `#rewind`, `#size` and `#close`. If the
+    # object doesn't respond to one of these methods, a Shrine::InvalidFile
+    # error is raised.
+    def _enforce_io(io)
+      missing_methods = %i[read eof? rewind close].select { |m| !io.respond_to?(m) }
+      raise InvalidFile.new(io, missing_methods) if missing_methods.any?
+    end
 
-        # Returns an opened IO object for the uploaded file by calling `#open`
-        # on the storage.
-        def io
-          @io || open
-        end
-      end
+    # Generates a unique identifier that can be used for a location.
+    def generate_uid(io)
+      SecureRandom.hex
     end
   end
+end
+
+require 'shrine/plugins'
+require 'shrine/attacher'
+require 'shrine/attachment'
+require 'shrine/uploaded_file'
```

Comment:
Normally I don't mind that files are required elsewhere in the file when needed, but here I don't see any reason why we wouldn't require them on the top. I think it's nice for the reader to see this on the top of the file.

Also, Shrine uses double quotes everywhere, even for `require` statements.